// @tanstack/router-core/dist/esm/index.js
import { TSR_DEFERRED_PROMISE, defer } from "./defer.js";
import { preloadWarning } from "./link.js";
import { isMatch } from "./Matches.js";
import { cleanPath, exactPathTest, interpolatePath, joinPaths, removeTrailingSlash, resolvePath, trimPath, trimPathLeft, trimPathRight } from "./path.js";
import { decode, encode } from "./qss.js";
import { rootRouteId } from "./root.js";
import { BaseRootRoute, BaseRoute, BaseRouteApi } from "./route.js";
import { PathParamError, RouterCore, SearchParamError, defaultSerializeError, getInitialRouterState, getLocationChangeInfo, getMatchedRoutes, lazyFn, trailingSlashOptions } from "./router.js";
import { createRouterConfig } from "./config.js";
import { retainSearchParams, stripSearchParams } from "./searchMiddleware.js";
import { defaultParseSearch, defaultStringifySearch, parseSearchWith, stringifySearchWith } from "./searchParams.js";
import { DEFAULT_PROTOCOL_ALLOWLIST, buildDevStylesUrl, createControlledPromise, deepEqual, escapeHtml, functionalUpdate, isDangerousProtocol, isModuleNotFoundError, isPlainArray, isPlainObject, replaceEqualDeep } from "./utils.js";
import { isRedirect, isResolvedRedirect, parseRedirect, redirect } from "./redirect.js";
import { isNotFound, notFound } from "./not-found.js";
import { defaultGetScrollRestorationKey, getCssSelector, handleHashScroll, restoreScroll, scrollRestorationCache, setupScrollRestoration, storageKey } from "./scroll-restoration.js";
import { createSerializationAdapter, makeSerovalPlugin, makeSsrSerovalPlugin } from "./ssr/serializer/transformer.js";
import { defaultSerovalPlugins } from "./ssr/serializer/seroval-plugins.js";
import { RawStream, createRawStreamDeserializePlugin, createRawStreamRPCPlugin } from "./ssr/serializer/RawStream.js";
import { composeRewrites, executeRewriteInput } from "./rewrite.js";
export {
  BaseRootRoute,
  BaseRoute,
  BaseRouteApi,
  DEFAULT_PROTOCOL_ALLOWLIST,
  PathParamError,
  RawStream,
  RouterCore,
  SearchParamError,
  TSR_DEFERRED_PROMISE,
  buildDevStylesUrl,
  cleanPath,
  composeRewrites,
  createControlledPromise,
  createRawStreamDeserializePlugin,
  createRawStreamRPCPlugin,
  createRouterConfig,
  createSerializationAdapter,
  decode,
  deepEqual,
  defaultGetScrollRestorationKey,
  defaultParseSearch,
  defaultSerializeError,
  defaultSerovalPlugins,
  defaultStringifySearch,
  defer,
  encode,
  escapeHtml,
  exactPathTest,
  executeRewriteInput,
  functionalUpdate,
  getCssSelector,
  getInitialRouterState,
  getLocationChangeInfo,
  getMatchedRoutes,
  handleHashScroll,
  interpolatePath,
  isDangerousProtocol,
  isMatch,
  isModuleNotFoundError,
  isNotFound,
  isPlainArray,
  isPlainObject,
  isRedirect,
  isResolvedRedirect,
  joinPaths,
  lazyFn,
  makeSerovalPlugin,
  makeSsrSerovalPlugin,
  notFound,
  parseRedirect,
  parseSearchWith,
  preloadWarning,
  redirect,
  removeTrailingSlash,
  replaceEqualDeep,
  resolvePath,
  restoreScroll,
  retainSearchParams,
  rootRouteId,
  scrollRestorationCache,
  setupScrollRestoration,
  storageKey,
  stringifySearchWith,
  stripSearchParams,
  trailingSlashOptions,
  trimPath,
  trimPathLeft,
  trimPathRight
};
